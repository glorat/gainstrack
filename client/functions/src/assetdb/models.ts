../../../src/lib/assetdb/models.ts