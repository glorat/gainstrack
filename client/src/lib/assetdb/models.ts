export interface AccountCreation {
    date: string;
    key: AccountKey;
    options: AccountOptions;
}

export interface AccountDTO {
    date: string
    accountId: string
    ccy: string
    options: AccountOptions
}

export interface AccountKey {
    name: string;
    assetId: string;
}

export interface AccountOptions {
    tradingAccount: boolean
    fundingAccount: string
    autoReinvestment: boolean
    multiAsset: boolean
    generatedAccount: boolean
}

export interface Amount {
    number: number
    ccy: string
}

export interface AmountEditing {
  number: number|undefined
  ccy: string|undefined
}

export interface QuoteConfig {
    avSymbol: string
    actualCcy: string
}

export interface AccountDTO {
    accountId: string
    date: string
    ccy: string
}

export interface AccountCommandDTO {
    accountId: string
    date: string
    asset?: string
    change?: Amount
    balance?: Amount
    price?: Amount
    otherAccount?: string
    options?: Record<string, any>
    commandType?: string
    // TODO: Put this in options or make it official
    commission?: Amount
  description?: string
}

export interface AccountCommandEditing {
  accountId: string // TODO: Make this defaultable too one day?
  date?: string
  asset?: string
  change?: AmountEditing
  balance?: AmountEditing
  price?: AmountEditing
  otherAccount?: string
  options?: Record<string, any>
  commandType?: string
  // TODO: Put this in options or make it official
  commission?: AmountEditing
  description?: string
}


export interface AuthenticationDTO {
    username: string
}


export interface NetworthByAsset {
    assetId: string
    value: number
    units: number
    price: number
    priceDate?: string
    priceMoves: Record<string, number>
}

export interface AssetColumn {
    name: string,
    label: string | number
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    field: string | ((row: any) => any)
    classes?: string[]
    align?: string
    sortable?: boolean
    format?: (val: number) => string
    tag?: string
    value?: string
}

export interface AssetResponse {
    rows: NetworthByAsset[]
    columns: Array<Record<string, any>>
    totals: NetworthByAsset[]
}

export interface Posting {
  account: string
  value: Amount
  price?: Amount
}

/** Denormalised posting view that includes Tx data*/
export interface PostingEx extends Posting {
  date: string
  originIndex: number
}

export interface HasPostings {
  postings: Posting[]
}

export interface Transaction extends HasPostings {
    originIndex: number
    id: number
    postDate: string
    postings: Posting[]
}



export interface AccountState {
    accounts: AccountCommandDTO[]
    baseCurrency: string
}

export interface AssetOptions extends Record<string, any> {
    tags: string[]
    ticker?: string
    proxy?: string
}

export interface AssetDTO {
    // date: string,
    asset: string,
    options: AssetOptions
}

export interface BalanceStateSeries {
    series: Record<string, number>
    ccy: string
}

type LocalDate = string

export interface PriceState {
    ccys: string[]
    // key: assetPair
    prices: Record<string, Record<LocalDate, number>>
}

export interface StateSummaryDTO {
  baseCcy: string;
  accounts: AccountDTO[]
  ccys: string[]
  authentication: AuthenticationDTO
  commands: AccountCommandDTO[]
  customToken?: string
}

export interface AllState extends StateSummaryDTO {
  accounts: AccountDTO[]
  commands: AccountCommandDTO[]
  assetState: AssetDTO[]
  balances: Record<string, BalanceStateSeries>
  txs: (Transaction | AccountCommandDTO)[]
  priceState: PriceState
  tradeFx: { baseCcy: string, data: Record<string, { ks: string[], vs: number[] }> }
  fxMapper: Record<string, string>,
  proxyMapper: Record<string, string>,
}

export const emptyAllState: AllState = {
  accounts: [],
  authentication: {username: ''},
  baseCcy: '', ccys: [],
  commands: [],
  assetState: [],
  balances: {},
  txs: [],
  priceState: {ccys:[], prices:{} },
  tradeFx: {baseCcy: 'USD', data:{}},
  fxMapper: {},
  proxyMapper:{}
};

export interface TreeTableDTO {
  name: string
  shortName: string
  children: TreeTableDTO[]
  assetBalance: { ccy: string, number: number }[]
}

export function isTransaction(tx: AccountCommandDTO|Transaction):tx is Transaction {
  return (tx as Transaction).postings !== undefined;
}
