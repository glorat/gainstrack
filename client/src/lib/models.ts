export interface AccountCreation {
    date: string;
    key: AccountKey;
    options: AccountOptions;
}

export interface AccountDTO {
    date: string
    accountId: string
    ccy: string
    options: AccountOptions
}

export interface AccountKey {
    name: string;
    assetId: string;
}

export interface AccountOptions {
    tradingAccount: boolean
    fundingAccount: string
    autoReinvestment: boolean
    multiAsset: boolean
    generatedAccount: boolean
}

export interface Amount {
    number: number
    ccy: string
}

export interface QuoteConfig {
    avSymbol: string
    actualCcy: string
}

export interface AccountDTO {
    accountId: string
    date: string
    ccy: string
}

export interface AccountCommandDTO {
    accountId: string
    date: string
    asset?: string
    change?: Amount
    balance?: Amount
    price?: Amount
    otherAccount?: string
    options?: Record<string, any>
    commandType?: string
    // TODO: Put this in options or make it official
    commission?: Amount
  description?: string
}

export interface AuthenticationDTO {
    username: string
}


export interface NetworthByAsset {
    assetId: string
    value: number
    price: number
    priceDate?: string
    priceMoves: Record<string, number>
}

export interface AssetColumn {
    name: string,
    label: string | number
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    field: string | ((row: any) => any)
    classes?: string[]
    align?: string
    sortable?: boolean
    format?: (val: number) => string
    tag?: string
    value?: string
}

export interface AssetResponse {
    rows: NetworthByAsset[]
    columns: Array<Record<string, any>>
    totals: NetworthByAsset[]
}

export interface Posting {
  account: string
  value: Amount
  price?: Amount
}

/** Denormalised posting view that includes Tx data*/
export interface PostingEx extends Posting {
  date: string
  originIndex: number
}

export interface HasPostings {
  postings: Posting[]
}

export interface Transaction extends HasPostings {
    originIndex: number
    id: number
    postDate: string
    postings: Posting[]
}



export interface AccountState {
    accounts: AccountCommandDTO[]
    baseCurrency: string
}

export interface AssetOptions extends Record<string, any> {
    tags: string[]
    ticker?: string
    proxy?: string
}

export interface AssetDTO {
    // date: string,
    asset: string,
    options: AssetOptions
}

export interface AssetState {
    allAssets: AssetDTO[]
    assetToTags: Record<string, string[]>
    tagToAssets: Record<string, string[]>
}

export interface BalanceStateSeries {
    series: Record<string, number>
    ccy: string
}

type LocalDate = string

export interface PriceState {
    ccys: string[]
    // key: assetPair
    prices: Record<string, Record<LocalDate, number>>
}

export interface StateSummaryDTO {
  baseCcy: string;
  accounts: AccountDTO[]
  ccys: string[]
  authentication: AuthenticationDTO
  commands: AccountCommandDTO[]
  customToken?: string
}

export interface AllState extends StateSummaryDTO {
  accounts: AccountDTO[]
  commands: AccountCommandDTO[]
  assetState: AssetState
  balances: Record<string, BalanceStateSeries>
  txs: (Transaction | AccountCommandDTO)[]
  priceState: PriceState
  tradeFx: { baseCcy: string, data: Record<string, { ks: string[], vs: number[] }> }
  fxMapper: Record<string, string>,
  proxyMapper: Record<string, string>,
}

export const emptyAllState: AllState = {
  accounts: [],
  authentication: {username: ''},
  baseCcy: '', ccys: [],
  commands: [],
  assetState: {allAssets:[], assetToTags:{}, tagToAssets:{}},
  balances: {},
  txs: [],
  priceState: {ccys:[], prices:{} },
  tradeFx: {baseCcy: 'USD', data:{}},
  fxMapper: {},
  proxyMapper:{}
};

export interface TreeTableDTO {
  name: string
  shortName: string
  children: TreeTableDTO[]
  assetBalance: { ccy: string, number: number }[]
}

export function isTransaction(tx: AccountCommandDTO|Transaction):tx is Transaction {
  return (tx as Transaction).postings !== undefined;
}
