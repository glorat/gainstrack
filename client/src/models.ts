export interface AccountCreation {
    date: string;
    key: AccountKey;
    options: AccountOptions;
}

export interface AccountDTO {
    date: string
    accountId: string
    ccy: string
    options: AccountOptions
}

export interface AccountKey {
    name: string;
    assetId: string;
}

export interface AccountOptions {
    tradingAccount: boolean
    fundingAccount: string
    autoReinvestment: boolean
    multiAsset: boolean
    generatedAccount: boolean
}

export interface Amount {
    number: number
    ccy: string
}

export interface QuoteConfig {
    avSymbol: string
    actualCcy: string
}

export interface AccountDTO {
    accountId: string
    date: string
    ccy: string
}

export interface AccountCommandDTO {
    accountId: string
    date: string
    change?: Amount
    balance?: Amount
    price?: Amount
    otherAccount?: string
    options?: Record<string, any>
    commandType?: string
    // TODO: Put this in options or make it official
    commission?: Amount
}

export interface StateSummaryDTO {
    accountIds: string[]
    accounts: AccountDTO[]
    ccys: string[]
    authentication: AuthenticationDTO
    commands: AccountCommandDTO[]
}

export interface AuthenticationDTO {
    username: string
}
