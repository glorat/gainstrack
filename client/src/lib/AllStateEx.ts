import {AccountCommandDTO, AccountDTO, AllState, isTransaction, Posting, PostingEx, Transaction} from './assetdb/models';
import {flatten} from 'lodash';
import {SingleFXConversion} from '../lib/fx';

export class AllStateEx {

  constructor(readonly state: AllState) {
  }

  allTxs() {
    return this.state.txs.filter(isTransaction);
  }

  allPostings() {
    const allTxPostings : Posting[][] = this.state.txs.map(tx => isTransaction(tx) ?  tx.postings : []);
    const allPostings = flatten(allTxPostings);
    return allPostings;
  }

  allPostingsEx() {
    const ret:PostingEx[] = [];
    const allTxs: Transaction[] = this.allTxs();
    allTxs.forEach(tx => {
      tx.postings.forEach(p => {
        ret.push({...p, date: tx.postDate, originIndex: tx.originIndex})
      });
    });
    return ret;
  }

  tradeFxConverter() {
    const tradeFxData: { baseCcy: string; data: Record<string, { ks: string[]; vs: number[] }> } | undefined = this.state.tradeFx;
    return SingleFXConversion.fromDTO(tradeFxData.data, tradeFxData.baseCcy)
  }

  underlyingCcy(assetId: string, accountId: string|undefined): string|undefined {
    const cmds = this.state.commands;
    const unitFilter = (cmd:AccountCommandDTO) => cmd.commandType === 'unit' && cmd.balance?.ccy === assetId;
    const tradeFilter = (cmd:AccountCommandDTO) => cmd.commandType === 'trade' && cmd.change?.ccy === assetId;
    const acctFilter = (cmd:AccountCommandDTO) => cmd.accountId === accountId;
    const reversed = [...cmds].reverse(); // Don't mutate!
    let prev = reversed.find(cmd => acctFilter(cmd) && (unitFilter(cmd) || tradeFilter(cmd)));
    if (!prev) prev = reversed.find(cmd => (unitFilter(cmd) || tradeFilter(cmd)));
    if (prev && prev.commandType === 'unit') return prev.price?.ccy;
    if (prev && prev.commandType === 'trade') return prev.price?.ccy;
    return undefined;
  }

  findAccount(accountId: string|undefined) {
    const all: AccountDTO[] = this.state.accounts;
    const acct = all.find(x => x.accountId === accountId);
    return acct
  }
}
