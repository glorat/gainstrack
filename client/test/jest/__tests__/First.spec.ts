import fs from 'fs';
import {AllState} from 'src/lib/models';
import {AllStateEx} from 'src/lib/AllStateEx';
import {isSubAccountOf, positionSetFx, positionUnderAccount, convertedPositionSet} from 'src/lib/utils';
import { LocalDate } from '@js-joda/core';
import {pnlExplain, pnlExplainMonthly} from 'src/lib/PLExplain';

describe('First', () => {
  describe('State', () => {
    const data = fs.readFileSync('test/jest/__tests__/first.json', 'utf8' );
    const allState: AllState = JSON.parse(data);
    const allStateEx = new AllStateEx(allState)
    const today = LocalDate.parse('2019-12-31');

    test('should parse test input', () => {
      expect(allState.baseCcy).toBe('GBP');
    })

    test('should correct balances', () => {
      const pSet = positionUnderAccount(allStateEx.allPostingsEx(), 'Assets:Investment:IBUSD:USD')
      expect(pSet['USD']).toBeCloseTo(172.05)

      const pSet2 = positionUnderAccount(allStateEx.allPostingsEx(), 'Expenses:Investment:IBUSD:USD')
      expect(pSet2['USD']).toBeCloseTo(18.87)

      const pSet3 = positionUnderAccount(allStateEx.allPostingsEx(), 'Assets')
      expect(pSet3['USD']).toBeCloseTo(-52857.23)
    });

    test('should have a list of all ccys', () => {
      expect(allState.ccys.length).toBe(18)
    });

    test('should convert balances', () => {
      const testMe = (accountId: string, strategy: string, ccy: string, expected: number) => {
        const pSet = positionUnderAccount(allStateEx.allPostingsEx(), accountId);
        const account = allState.accounts.find(x => x.accountId === accountId)
        // const value = positionSetFx(pSet, ccy, today, allStateEx.tradeFxConverter())
        const value = convertedPositionSet(pSet, allState.baseCcy, strategy, today, account, allStateEx.tradeFxConverter())
        expect(Math.round(value[ccy])).toBe(expected)
      }

      testMe('Assets:Investment:IBUSD:USD', 'parent', 'USD', 172)
      testMe('Expenses:Investment:IBUSD:USD', 'parent', 'USD', 19)
      testMe('Assets:Investment:IBUSD', 'parent', 'USD', 34960)

      testMe('Assets:Investment', 'parent', 'GBP', 433653)
      testMe('Assets', 'parent', 'GBP', 625582)


      testMe('Assets:Investment:IBUSD:USD', 'GBP', 'GBP', 135)
      testMe('Assets', 'GBP', 'GBP', 625582)

      testMe('Assets:Investment:IBUSD', 'units', 'USD', 172)

      testMe('Assets:Investment:IBUSD', 'USD', 'USD', 34960)
      testMe('Assets', 'USD', 'USD', 797742)
    });


    test('pnlExplain monthly', () => {
      const pnl = pnlExplainMonthly(today, allStateEx.allPostingsEx(), allState.commands, allState.baseCcy, allStateEx.tradeFxConverter());
      expect(pnl).toMatchSnapshot();
    })
  })

});
