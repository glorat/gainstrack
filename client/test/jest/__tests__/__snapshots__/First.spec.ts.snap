// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First State asset tree converted to global 1`] = `
Object {
  "assetBalance": Array [
    Object {
      "ccy": "GBP",
      "number": 625581.5365236884,
    },
  ],
  "children": Array [
    Object {
      "assetBalance": Array [
        Object {
          "ccy": "GBP",
          "number": 188635.59,
        },
      ],
      "children": Array [
        Object {
          "assetBalance": Array [
            Object {
              "ccy": "GBP",
              "number": 19000,
            },
          ],
          "children": Array [],
          "name": "Assets:Bank:HSBCUK",
          "shortName": "HSBCUK",
        },
        Object {
          "assetBalance": Array [
            Object {
              "ccy": "GBP",
              "number": 115703.48999999999,
            },
          ],
          "children": Array [],
          "name": "Assets:Bank:Nationwide",
          "shortName": "Nationwide",
        },
        Object {
          "assetBalance": Array [
            Object {
              "ccy": "GBP",
              "number": 53932.1,
            },
          ],
          "children": Array [],
          "name": "Assets:Bank:NationwideSavings",
          "shortName": "NationwideSavings",
        },
      ],
      "name": "Assets:Bank",
      "shortName": "Bank",
    },
    Object {
      "assetBalance": Array [
        Object {
          "ccy": "GBP",
          "number": 0,
        },
      ],
      "children": Array [],
      "name": "Assets:HSBCCN",
      "shortName": "HSBCCN",
    },
    Object {
      "assetBalance": Array [
        Object {
          "ccy": "GBP",
          "number": 3292.743399472632,
        },
      ],
      "children": Array [],
      "name": "Assets:HSBCHK",
      "shortName": "HSBCHK",
    },
    Object {
      "assetBalance": Array [
        Object {
          "ccy": "GBP",
          "number": 433653.2031242158,
        },
      ],
      "children": Array [
        Object {
          "assetBalance": Array [
            Object {
              "ccy": "GBP",
              "number": -22107.826223337514,
            },
          ],
          "children": Array [
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 2429.383626097867,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:HSBC:BRK-B",
              "shortName": "BRK-B",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": -41585.069008782935,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:HSBC:USD",
              "shortName": "USD",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 806.0774780426599,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:HSBC:VCSH",
              "shortName": "VCSH",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 16241.781681304896,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:HSBC:VTI",
              "shortName": "VTI",
            },
          ],
          "name": "Assets:Investment:HSBC",
          "shortName": "HSBC",
        },
        Object {
          "assetBalance": Array [
            Object {
              "ccy": "GBP",
              "number": 27415.40934755332,
            },
          ],
          "children": Array [
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 971.7534504391468,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:IBUSD:BRK-B",
              "shortName": "BRK-B",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 4097.396486825595,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:IBUSD:IUAA",
              "shortName": "IUAA",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 134.92001254704837,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:IBUSD:USD",
              "shortName": "USD",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 22211.33939774153,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:IBUSD:VWRD",
              "shortName": "VWRD",
            },
          ],
          "name": "Assets:Investment:IBUSD",
          "shortName": "IBUSD",
        },
        Object {
          "assetBalance": Array [
            Object {
              "ccy": "GBP",
              "number": 80300.28,
            },
          ],
          "children": Array [
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 0.0017671099985818728,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:GBP",
              "shortName": "GBP",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 1802.43824,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:GBR04S9K",
              "shortName": "GBR04S9K",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 3981.925561,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:GBR06HUW",
              "shortName": "GBR06HUW",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 18405.485914,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:OPXC",
              "shortName": "OPXC",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 17342.540658,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:OPY7",
              "shortName": "OPY7",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 3475.46032769,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:QLNC",
              "shortName": "QLNC",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 33385.518261,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:QS90",
              "shortName": "QS90",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 1906.9092712,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:OldMutual:SVDK",
              "shortName": "SVDK",
            },
          ],
          "name": "Assets:Investment:OldMutual",
          "shortName": "OldMutual",
        },
        Object {
          "assetBalance": Array [
            Object {
              "ccy": "GBP",
              "number": 348045.33999999997,
            },
          ],
          "children": Array [
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 155303.3314,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:Zurich:AMER",
              "shortName": "AMER",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 158121.6,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:Zurich:FTSE",
              "shortName": "FTSE",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": -0.020100000030652154,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:Zurich:GBP",
              "shortName": "GBP",
            },
            Object {
              "assetBalance": Array [
                Object {
                  "ccy": "GBP",
                  "number": 34620.4287,
                },
              ],
              "children": Array [],
              "name": "Assets:Investment:Zurich:TBOND",
              "shortName": "TBOND",
            },
          ],
          "name": "Assets:Investment:Zurich",
          "shortName": "Zurich",
        },
      ],
      "name": "Assets:Investment",
      "shortName": "Investment",
    },
  ],
  "name": "Assets",
  "shortName": "Assets",
}
`;

exports[`First State pnlExplain monthly 1`] = `
Array [
  Object {
    "actual": 13149.88062177109,
    "delta": Array [],
    "explained": 13149.880621771095,
    "fromDate": "2019-12-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 45.81369226032758,
    "tenor": "total",
    "toDate": "2019-01-31",
    "toNetworth": 0,
    "totalDeltaExplain": 65.42310719596667,
    "totalEquity": 13039.876962214425,
    "totalExpense": 1.2331398996235887,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": -4.198123330449259e-12,
  },
  Object {
    "actual": 157798.5674612531,
    "delta": Array [],
    "explained": 157798.56746125314,
    "fromDate": "2019-12-01",
    "networthChange": 0.33694631920989737,
    "newActivityByAccount": Array [],
    "newActivityPnl": 549.764307123931,
    "tenor": "total",
    "toDate": "2019-01-31",
    "toNetworth": 0,
    "totalDeltaExplain": 785.0772863515999,
    "totalEquity": 156478.5235465731,
    "totalExpense": 14.797678795483064,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": -5.03774799653911e-11,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-12-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Dec",
    "toDate": "2019-12-31",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-11-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Nov",
    "toDate": "2019-11-30",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-10-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Oct",
    "toDate": "2019-10-31",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-09-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Sep",
    "toDate": "2019-09-30",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-08-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Aug",
    "toDate": "2019-08-31",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-07-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Jul",
    "toDate": "2019-07-31",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-06-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Jun",
    "toDate": "2019-06-30",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 0,
    "delta": Array [],
    "explained": 0,
    "fromDate": "2019-05-01",
    "networthChange": 0,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "May",
    "toDate": "2019-05-31",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 0,
  },
  Object {
    "actual": 39.66436637379229,
    "delta": Array [
      Object {
        "amount": 15,
        "assetId": "BRK-B",
        "explain": 51.32528230865728,
        "fromDate": "2019-03-31",
        "newPrice": 161.95890840652447,
        "newValue": 2429.383626097867,
        "oldPrice": 158.53722291928065,
        "oldValue": 2378.0583437892096,
        "toDate": "2019-04-30",
      },
    ],
    "explained": 39.66436637389633,
    "fromDate": "2019-04-01",
    "networthChange": 0.00006340801174028729,
    "newActivityByAccount": Array [
      Object {
        "accountId": "BUY 100.0 VWRD @85.79 USD",
        "explain": 3.136762860726776,
      },
      Object {
        "accountId": "BUY 230.0 VWRD @85.83 USD",
        "explain": -3.637978807091713e-12,
      },
      Object {
        "accountId": "BUY 1000.0 IUAA @5.225 USD",
        "explain": -9.094947017729282e-13,
      },
      Object {
        "accountId": "BUY 6.0 BRK-B @206.53 USD",
        "explain": -1.1368683772161603e-13,
      },
    ],
    "newActivityPnl": 3.136762860722115,
    "tenor": "Apr",
    "toDate": "2019-04-30",
    "toNetworth": 625581.5365236884,
    "totalDeltaExplain": 51.32528230865728,
    "totalEquity": -0,
    "totalExpense": 14.797678795483064,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": -1.0403766736999387e-10,
  },
  Object {
    "actual": 163247.49036177335,
    "delta": Array [
      Object {
        "amount": 100,
        "assetId": "VTI",
        "explain": 215.61977206190204,
        "fromDate": "2019-02-28",
        "newPrice": 112.79015056461733,
        "newValue": 11279.015056461734,
        "oldPrice": 110.63395284399833,
        "oldValue": 11063.395284399832,
        "toDate": "2019-03-31",
      },
    ],
    "explained": 163247.49036177332,
    "fromDate": "2019-03-01",
    "networthChange": 0.3531245388008474,
    "newActivityByAccount": Array [
      Object {
        "accountId": "BUY 13.0 VCSH @79.07 USD",
        "explain": -2.2737367544323206e-13,
      },
      Object {
        "accountId": "BUY 14.0 VTI @144.62 USD",
        "explain": -8.673149309911878,
      },
      Object {
        "accountId": "BUY 30.0 VTI @143.83 USD",
        "explain": 9.094947017729282e-13,
      },
      Object {
        "accountId": "BUY 15.0 BRK-B @200.58 USD",
        "explain": 18.663739021330002,
      },
    ],
    "newActivityPnl": 9.990589711418806,
    "tenor": "Mar",
    "toDate": "2019-03-31",
    "toNetworth": 625541.8721573146,
    "totalDeltaExplain": 215.61977206190204,
    "totalEquity": 163021.88,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 2.9103830456733704e-11,
  },
  Object {
    "actual": -6025.224221445795,
    "delta": Array [
      Object {
        "amount": 100,
        "assetId": "VTI",
        "explain": 518.1322319810406,
        "fromDate": "2019-01-31",
        "newPrice": 110.63395284399833,
        "newValue": 11063.395284399832,
        "oldPrice": 105.45263052418792,
        "oldValue": 10545.263052418792,
        "toDate": "2019-02-28",
      },
    ],
    "explained": -6025.224221445872,
    "fromDate": "2019-02-01",
    "networthChange": -0.012865624552193628,
    "newActivityByAccount": Array [],
    "newActivityPnl": 0,
    "tenor": "Feb",
    "toDate": "2019-02-28",
    "toNetworth": 462294.3817955412,
    "totalDeltaExplain": 518.1322319810406,
    "totalEquity": -6543.356453426913,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": 7.73070496506989e-11,
  },
  Object {
    "actual": 536.6369545517373,
    "delta": Array [],
    "explained": 536.63695455179,
    "fromDate": "2019-01-01",
    "networthChange": 0.0011471921597045404,
    "newActivityByAccount": Array [
      Object {
        "accountId": "BUY 100.0 VTI @127.63 USD",
        "explain": 536.63695455179,
      },
    ],
    "newActivityPnl": 536.63695455179,
    "tenor": "Jan",
    "toDate": "2019-01-31",
    "toNetworth": 468319.606016987,
    "totalDeltaExplain": 0,
    "totalEquity": -0,
    "totalExpense": 0,
    "totalIncome": -0,
    "totalYieldIncome": 0,
    "unexplained": -5.275069270282984e-11,
  },
]
`;
